modules/gateway/peers.go

```diff
@@ -290,6 +458,9 @@ func (g *Gateway) Connect(addr modules.NetAddress) error {
 	if err := addr.IsValid(); err != nil {
 		return errors.New("can't connect to invalid address")
 	}
+	if net.ParseIP(addr.Host()) == nil {
+		return errors.New("address must be an IP address")
+	}
```

Comment:
There's no reason to support non-IP addresses that I can think of. Discovered hostnames can only be IP addresses (unless UPnP supports non-IP hostname discovery; I don't think it does).

It would be deceptive for `Connect` to allow a non-IP address, as the peer it is connecting to would only be aware of its IP address. It is this IP address that the receiving peer adds to its node list. That is, unless `PortHandshake` becomes `AddressHandshake` which isn't unreasonable. But then you would want to make sure that the address the peer claims is it's address matches the `conn.RemoteAddr()`, by doing a DNS lookup.

It's possible, but adds more complexity than I think is necessary. I don't see any advantage to supporting non-IP addresses, especially since there is no current mechanism to set a peer's hostname as there is in the host module.